fix(api): Addition of distribution error to prevent invalid disposal values

CaseyBatten opened the pull request:
# Overview
Prevents the distribute functionality from accepting a disposal value equal to or greater than the maximum value of the instrument. This was necessary to prevent an infinite looping issue where when disposal values greater than or equal to the maximum volume of an instrument were provided, which would cause the looping component of the _plan_distribute function to get caught in an state that was impossible to complete. This would seek to solve the issue detailed in RSS-347.

CaseyBatten committed:
Addition of error to prevent invalid disposal values

Prevents the distribute functionality from accepting a disposal value equal to or greater than the maximum value of the instrument.

ecormany commented on api/src/opentrons/protocols/advanced_control/transfers.py:
```diff
@@ -571,6 +571,12 @@ def _plan_distribute(self):
+
+        if self._strategy.disposal_volume >= self._instr.max_volume:
+            raise ValueError(
+                "The Disposal Volume must be less than the Maximum Volume of the Instrument"
```
Looks good! We can probably lowercase all these terms in the error text:

```suggestion
                "The disposal volume must be less than the maximum volume of the pipette"
```

CaseyBatten committed:
Message update and _plan_distribute simplified

Updated error message to be consistent with desired message, moved error raise to a validation function call to ensure _plan_distribute is not too complex

SyntaxColoring commented on api/src/opentrons/protocols/advanced_control/transfers.py:
```diff
@@ -534,6 +534,12 @@ def _extend_source_target_lists(
         return sources, targets
 
+    def _check_valid_disposal_volume(self):
+        if self._strategy.disposal_volume >= self._instr.max_volume:
```
* Do we need to check for the "working volume" or something like that, instead of `self._instr.max_volume` here?
* Do we need to check for negative numbers?

ecormany replied:
Not for fixing the narrow bug: infinite loop on analysis. It has to be ≥ the pipette max volume to trigger that.

You can get into other (less bad? merely different?) bug behavior by doing things like having a Flex 1000 µL pipette pick up a 50 µL tip and then specifying disposal volume of ≥50. The whole `distribute()` command no-ops then. The same goes for 200 µL filter tips on a 300 µL OT-2 pipette, etc. See RSS-344.

SyntaxColoring replied:
Good grief.

SyntaxColoring replied:
See https://github.com/Opentrons/opentrons/pull/13659#issuecomment-1738022236.

SyntaxColoring commented on api/src/opentrons/protocols/advanced_control/transfers.py:
```diff
@@ -571,6 +577,9 @@ def _plan_distribute(self):
 
         """
+
+        self._check_valid_disposal_volume()
```
Let's add tests for this. The unit tests in `test_transfers.py` are probably the right place to start, unless those turn out to be horrific.

SyntaxColoring reviewed (commented):
Looks great, thank you! A couple notes for moving forward with this.

SyntaxColoring commented:
Okay, I've done some archeology and found that this was previously reported as GitHub issue #6170 and we tried to fix it in PR #6754. We closed that PR without merging it because of [these reasons](https://github.com/Opentrons/opentrons/pull/6754#issuecomment-804499335). 2.5 years later, I no longer think those reasons are good. In particular, screw this:

> ...by my understanding of our [Python Protocol API's versioning policy](https://docs.opentrons.com/v2/versioning.html), we must characterize and preserve that behavior in fixing these bugs. Even though it was undocumented, and didn't make sense, and only happened in incorrectly written protocols.

Versioning in the Python Protocol API is good to avoid flagrantly breaking people's protocols. But it shouldn't prevent us from fixing flagrant *bugs* like infinite loops. That's not helping anybody.

-----

So, for this PR, here's what we should do:

1. Let's definitely add tests. Let's start with the ones from PR #6754. We'll have to lightly adapt them to work in our modern codebase, and we'll 
2. As described in [the old issue](https://github.com/Opentrons/opentrons/issues/6170), we should solve this for `air_gap` *and* `disposal_volume`.

CaseyBatten committed:
Adjustment to working volume from max and test case

Working volume used to better reflect operating case of distribute, test case added to test transfers regarding distribute test

CaseyBatten committed:
Addition of air gap to the error throw, expansion of tests cases

Test cases added/modified for disposal vol and air gaps, addition of calls to check for this issue across all transfer plans, assertion of check in validation

SyntaxColoring commented on api/src/opentrons/protocols/advanced_control/transfers.py:
```diff
@@ -484,6 +484,11 @@ def _plan_transfer(self):
+            self._strategy.air_gap,
+            self._instr._core.get_working_volume(),
+        )
         sources, dests = self._extend_source_target_lists(self._sources, self._dests)
```
A few things here:

1. For readability, let's move this `_check_valid_volume_parameters()` call below the `self._extend_source_target_lists()` call. We want to avoid interrupting the `# reform source target lists` comment, and we also want to keep the `_check_valid_volume_parameters()` call as close as possible to the `self._instr.max_volume - self._strategy.disposal_volume - self._strategy.air_gap,max_volume` calculation.
2. This is a nitpicky style thing, but I suggest providing the arguments as keywords, instead of positionally:

   ```python
    self._check_valid_volume_parameters(
        disposal_volume=self._strategy.disposal_volume,
        air_gap=self._strategy.air_gap,
        max_volume=self._instr._core.get_working_volume(),
    )
   ```

   We have 3 arguments and they're all floats, so I'm worried about them getting mixed up.
3. This particular call to `_check_valid_volume_parameters()` should use *`self._instr._core.max_volume`,* not `self._instr._core.get_working_volume()`. We need to exactly match what's in the call to `self._expand_for_volume_constraints()`.

   To make this kind of mistake harder to make, we could do the calculation and validation in one place. [Something like this.](https://gist.github.com/SyntaxColoring/c88f9c2d5b97ecf1dd2ab3cfc6f17c1c)

SyntaxColoring commented on api/src/opentrons/protocols/advanced_control/transfers.py:
```diff
@@ -633,6 +645,7 @@ def _expand_for_volume_constraints(
         """Split a sequence of proposed transfers if necessary to keep each
         transfer under the given max volume.
         """
+        assert max_volume > 0
```
Let's leave a comment here to explain the intent:

```suggestion
        # A final defense against an infinite loop.
        # Raising a proper exception with a helpful message is left to calling code,
        # because it has more context about what the user is trying to do.
        assert max_volume > 0
```

SyntaxColoring commented on api/src/opentrons/protocols/advanced_control/transfers.py:
```diff
@@ -571,6 +576,13 @@ def _plan_distribute(self):
+            self._strategy.air_gap,
+            self._instr._core.get_working_volume(),
+        )
+
```
Ditto my previous comments about:

1. Providing arguments with keyword syntax instead of positionally.
2. Using `self._instr.max_volume` instead of `self._instr._core.get_working_volume()`. And maybe combining the calculation with the validation to make that kind of mistake harder to make.

SyntaxColoring commented on api/src/opentrons/protocols/advanced_control/transfers.py:
```diff
@@ -680,6 +693,11 @@ def _plan_consolidate(self):
+            self._strategy.disposal_volume,
+            self._strategy.air_gap,
+            self._instr._core.get_working_volume(),
+        )
```
Ditto my previous comment about providing arguments with keyword syntax, instead of positionally.

SyntaxColoring commented on api/src/opentrons/protocols/advanced_control/transfers.py:
```diff
@@ -843,6 +861,20 @@ def _map_volume(i):
 
         return [_map_volume(i) for i in range(total)]
 
+    def _check_valid_volume_parameters(self, disposal_volume, air_gap, working_volume):
```
Let's give type annotations to this new function.

```suggestion
    def _check_valid_volume_parameters(self, disposal_volume: float, air_gap: float, working_volume: float) -> None:
```

Normally, forgetting the type annotations would be an error because of [our project config](https://github.com/Opentrons/opentrons/blob/c0543a6e0a89e38a06117c5124085a4bb45443c4/api/mypy.ini#L5), but I think we have that overridden for this file because it's legacy code and we haven't gotten around to fixing it up. So we *could* leave them out. But let's leave a little less work for our future selves and add them now.

SyntaxColoring commented on api/tests/opentrons/protocols/advanced_control/test_transfers.py:
```diff
@@ -796,6 +796,83 @@ def test_all_options(_instr_labware):
+            mode="distribute",
+            options=options,
+        )
+        list(plan)
```
This iterator thing was a little tricky, so let's leave an explanatory comment. Shamelessly copy-pasting from #6754:

```suggestion
        # Exhaust the iterator in case it raises the expected exception lazily.
        list(plan)
```

And ditto for the other two test functions.

SyntaxColoring commented on api/tests/opentrons/protocols/advanced_control/test_transfers.py:
```diff
@@ -796,6 +796,83 @@ def test_all_options(_instr_labware):
+    lw2 = _instr_labware["lw2"]
+    # Supply the disposal volume assessment with the instrument followed by a disposal value equal to the max volume
+
+    options = tx.TransferOptions()
+    options = options._replace(
+        transfer=options.transfer._replace(air_gap=_instr_labware["instr"].max_volume)
+    )
```
This creates a `tx.TransferOptions` with all default values, and then, as a second step, modifies `.transfer.air_gap` to be `._instr_labware["instr"].max_volume`.

That works, but it's more common to do it all in one step. That looks like this:

```suggestion
    options = tx.TransferOptions(
        transfer=tx.Transfer(
            air_gap=_instr_labware["instr"].max_volume
        )
    )
```

Ditto for the other two test functions.

SyntaxColoring commented on api/tests/opentrons/protocols/advanced_control/test_transfers.py:
```diff
@@ -796,6 +796,83 @@ def test_all_options(_instr_labware):
     assert xfer_plan_list == exp1
 
 
+def test_invalid_air_gap_disposal_sum_distribute(_instr_labware):
```
**I think we should leave this to a separate PR, but:**

This adds a few test functions that differ only in the values that they provide for `disposal_volume` and `air_gap`. This would be a good use case for [pytest parametrization](https://docs.pytest.org/en/7.1.x/how-to/parametrize.html#parametrize-basics). We could write this as one function that gets parametrized with the different `disposal_volume` and `air_gap` combinations.

We could also parametrize this over the *mode* of `TransferPlan`: `"distribute"`, `"consolidate"`, or `"transfer"`. That would be good because our bugfix *does* touch those other functions.

SyntaxColoring reviewed (commented):
This looks great, thank you. Here are several minor comments, and then this looks good to merge.

CaseyBatten committed:
consolidation of changes and improvements to test cases

Adjustment of calls, change over to max volume control, formatting and improvement of test case formatting

SyntaxColoring commented on api/src/opentrons/protocols/advanced_control/transfers.py:
```diff
@@ -680,6 +693,11 @@ def _plan_consolidate(self):
+            self._strategy.disposal_volume,
+            self._strategy.air_gap,
+            self._instr._core.get_working_volume(),
+        )
```
I just noticed the #todo comment right below here that points out how `air_gap` and `disposal_volume` aren't factoring into the loop *at all* here in `_plan_consolidate()`. That seems really suspicious to me, but for our sanity, I think we need to treat that as a separate bug for a separate PR. For now, let's remove `_check_valid_volume_parameters()` from `_plan_consolidate()`. We'll keep your fix scoped to the one infinite loop that we know and understand.

CaseyBatten replied:
Gotcha! Will do, gonna recommit with the plan_consolidate call to the volume param check removed.

CaseyBatten committed:
Disable validation call under _plan_consolidate

_plan_consolidate volume parameter check disabled

CaseyBatten committed:
Message formatting consistency

CaseyBatten commented:
Primary fail issues cleared, formatting good, I believe this is ready for merge.

SyntaxColoring reviewed (commented):
This looks good to merge, assuming CI passes. I'm a little worried about that failing G-code test. I'm checking now to see if it's something that was introduced by this PR, or whether it's the known issue that was fixed in https://github.com/Opentrons/opentrons/pull/13670.

SyntaxColoring reviewed (approved):
G-code tests are failing on `edge`, so I'm presuming those failures are unrelated to this PR. (I'm running into obscure unrelated errors trying to run them locally, so I can't be sure.)

This looks good to merge. Thank you!
